src/make/Processor.cpp

```diff
@@ -944,80 +960,257 @@ Processor::_BuildCommand(data::RuleActionsCall* actionsCall)
-				needsExpansion |= remainder + 1 < end && remainder[1] == '(';
+		// skip whitespace
+		if (std::isspace(*remainder)) {
+			if (wordStart != remainder) {
+				words.push_back({wordStart, remainder});
+			}
+			remainder++;
+			wordStart = remainder;
+		} else {
 			remainder++;
 		}
+	}
 
-		// append the sequence, expanding variables, if necessary
-		if (needsExpansion) {
-			// Check for empty expansions
-			const auto isVar = [wordStart](char varName) {
-				return wordStart[2] == varName;
-			};
-			const bool varIsTarget = isVar('1') || isVar('<');
-			const bool varIsSource = isVar('2') || isVar('>');
-
-			// If sources are being expanded and have been trimmed to empty by
-			// EXISTING or UPDATED, cancel this action.
-			if (varIsSource && sourcesEmpty
-				&& (isExistingAction || isUpdatedAction)) {
-				canceled = true;
-				break;
-			}
+	data::StringListList sources{};
+	if (actions->IsPiecemeal() && !boundSourceTargets.IsEmpty()) {
+		std::uint32_t maxLine =
+			actions->MaxLine() > 0 ? actions->MaxLine() : kMaxCommandLength;
 
-			// If a target list is otherwise empty and used, print a warning.
-			if ((sourcesEmpty && varIsSource)
-				|| (targetsEmpty && varIsTarget)) {
-				std::stringstream warning{};
-				warning << "action " << actionsCall->Actions()->RuleName()
-						<< " called with no ";
-				warning << (varIsSource ? "sources" : "targets");
-				_PrintWarning(warning.str());
-			}
+		sources = _PiecemealWords(
+			&fEvaluationContext,
+			words,
+			boundSourceTargets,
+			maxLine
+		);
+	} else {
+		sources.push_back(boundSourceTargets);
+	}
 
-			StringList result = code::Leaf::EvaluateString(
+	for (StringList commandSources : sources) {
+		data::VariableDomain builtInWithSources{builtInVariables};
+		builtInWithSources.Set("2", commandSources);
+		builtInWithSources.Set(">", commandSources);
+		fEvaluationContext.SetBuiltInVariables(&builtInWithSources);
+
+		// Build command
+		data::String commandLine{};
+		for (auto& [wordStart, wordEnd] : words) {
+			auto evaluatedWord = code::Leaf::EvaluateString(
 				fEvaluationContext,
 				wordStart,
-				remainder,
+				wordEnd,
 				nullptr
 			);
-
-			bool isFirst = true;
-			for (StringList::Iterator it = result.GetIterator();
-				 it.HasNext();) {
-				if (isFirst)
-					isFirst = false;
-				else
-					commandLine += ' ';
-				commandLine += it.Next();
-			}
-		} else
-			commandLine.Append(wordStart, remainder - wordStart);
+			const StringPart separator{" "};
+			commandLine = commandLine + evaluatedWord.Join(separator);
+		}
+		commands.push_back(new Command(actionsCall, commandLine, boundTargets));
 	}
 
 	// reinstate the old local variable scope and the built-in variables
 	fEvaluationContext.SetLocalScope(oldLocalScope);
 	fEvaluationContext.SetBuiltInVariables(oldBuiltInVariables);
-	return canceled ? nullptr
-					: new Command(actionsCall, commandLine, boundTargets);
+	return;
+}
+
+data::StringListList
+Processor::_PiecemealWords(
```

Comment:
Not sure if this fits Processor. Would move to some utility class.

Reply:
It's only useful in the context of _BuildCommands, so I'm not sure where else would be a good place to extract this.